File: C++/Solution130.cpp
#include<iostream>
#include<vector>

#include<unordered_map>
#include<unordered_set>

using namespace std;

class Solution130 {
public:
    int n, m;
    unordered_set<int> set;
    int xx[4] = {-1, 1, 0, 0};
    int yy[4] = {0, 0, -1, 1};

    int convert(int row, int col){
        return row * m + col;
    }

    void dfs(vector<vector<char>>& board, int x, int y){
        set.insert(convert(x, y));
        for(int i = 0; i < 4; i++){
            int curx = x + xx[i], cury = y + yy[i];
            if(curx >= 0 && curx < n && cury >= 0 && cury < m && board[curx][cury] == 'O' && set.count(convert(curx, cury)) == 0){
                dfs(board, curx, cury);
            }
        }
        
    }

    void solve(vector<vector<char>>& board) {
        n = board.size();
        m = board[0].size();
        for(int i = 0; i < n; i++){
            if(board[i][0] == 'O'){
                dfs(board, i, 0);
            }
            if(board[i][m - 1] == 'O'){
                dfs(board, i, m - 1);
            }
        }
        for(int i = 0; i < m; i++){
            if(board[0][i] == 'O'){
                dfs(board, 0, i);
            }
            if(board[n - 1][i] == 'O'){
                dfs(board, n - 1, i);
            }
        }
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(board[i][j] == 'O' && set.count(convert(i, j)) == 0){
                    board[i][j] = 'X';
                }
            }
        }

    }
};

File: Java/other/Solution130.java
import java.util.Set;
import java.util.HashSet;

class Solution130 {
	Set<Integer> set = new HashSet<>();
	int[] x = new int[]{-1, 1, 0, 0};
	int[] y = new int[]{0, 0, -1, 1};

    public void solve(char[][] board) {
        if(board.length == 0){
            return;
        }
    	for(int row = 0; row < board.length; row++){
    		if(board[row][0] == 'O'){
    			dfs(board, row, 0);
    		}
    		if(board[row][board[0].length - 1] == 'O'){
    			dfs(board, row, board[0].length - 1);
    		}
    	}

    	for(int col = 0; col < board[0].length; col++){
    		if(board[0][col] == 'O'){
    			dfs(board, 0, col);
    		}
    		if(board[board.length - 1][col] == 'O'){
    			dfs(board, board.length - 1, col);
    		}
    	}

    	for(int i = 0; i < board.length; i++){
    		for(int j = 0; j < board[0].length; j++){
    			if(board[i][j] == 'O' && !set.contains(i * board[0].length + j)){
    				board[i][j] = 'X';
    			}
    		}
    	}	
    }

    private void dfs(char[][] board, int row, int col){
    	set.add(row * board[0].length + col);
    	for(int index = 0; index < 4; index++){
    		if(row + x[index] >= 0 && row + x[index] < board.length && col + y[index] >= 0 && col + y[index] < board[0].length && board[row + x[index]][col + y[index]] == 'O'){
    			int cur = (row + x[index]) * board[0].length + col + y[index];
    			if(!set.contains(cur)){
    				dfs(board, row + x[index], col + y[index]);
    			}
    		}
    	}
    }

    public static void main(String[] args){
    	Solution130 s = new Solution130();
    	char[][] board = new char[][]{new char[]{'X', 'X', 'X', 'X'}, new char[]{'X', 'O', 'O', 'X'}, new char[]{'X', 'X', 'O', 'X'}, new char[]{'X', 'O', 'X', 'X'}};
    	s.solve(board);
    }
}